File: src/java/magma/collect/map/Map_.java
package magma.collect.map;

import magma.collect.stream.Stream;
import magma.option.Option;
import magma.option.Tuple;

import java.util.function.Function;
import java.util.function.Supplier;

public interface Map_<K, V> {
    Map_<K, V> with(K key, V value);

    Option<V> find(K key);

    Stream<Tuple<K, V>> stream();

    Map_<K, V> ensure(K key, Function<V, V> whenPresent, Supplier<V> whenEmpty);

    Map_<K, V> withAll(Map_<K, V> other);

    boolean isEmpty();

    Map_<K, V> remove(K key);

    boolean containsKey(K key);

    Stream<V> streamValues();

    Stream<K> streamKeys();
}

File: src/clang/magma/collect/map/Map_.c
#include "Map_.h"
magma.collect.map.Map_<K, V> with(K key, V value);
magma.option.Option<V> find(K key);
magma.collect.stream.Stream<magma.option.Tuple<K, V>> stream();
magma.collect.map.Map_<K, V> ensure(K key, V(*whenPresent)(V), V(*whenEmpty)());
magma.collect.map.Map_<K, V> withAll(magma.collect.map.Map_<K, V> other);
magma.collect.map.boolean isEmpty();
magma.collect.map.Map_<K, V> remove(K key);
magma.collect.map.boolean containsKey(K key);
magma.collect.stream.Stream<V> streamValues();

